File: Cpp_ws/ProgettoCpp/src/Animale.cpp
#include "Animale.h"

namespace animali {

int Animale::count = 0; //inizializzazioe variabile statica

Animale::Animale(float peso, bool genere, int eta): id("A" + to_string(++count)), eta(eta), peso(peso), maschio(genere), vivo(true) { }

Animale::~Animale() { }

const string& Animale::getId() const { return id; }

int Animale::getEta() const { return eta; }

void Animale::invecchia() { eta++;  }

const bool Animale::isMaschio() const { return maschio; }

float Animale::getPeso() const { return peso; }

void Animale::setPeso(float p) { peso = p; }

bool Animale::isVivo() const { return vivo; }

void Animale::setVivo(bool vivo) { this->vivo = vivo; }

string Animale::toString() {
	string s =  id + "\n";
	s += "   mesi = " + to_string(eta) + "\n";
	s += "   peso = " + to_string(peso) + " kg\n";
	s += "   sesso = " + string(maschio? "maschio\n" : "femmina\n");
	s += "   vivo = " + string(vivo? "sì\n" : "no\n");
	return s;
}

} /* namespace animali */

File: Java_ws/ProgettoJava/src/animali/Animale.java
package animali;

import riserva.BadInputException;

public abstract class Animale implements Comparable<Animale> {
	private static int count = 0; //campo statico per generare l'id
	
	//CAMPI
	protected final String id;
	protected int eta; //età dell'animale in mesi
	protected float peso;
	protected final Genere sesso;
	protected boolean vivo;

	//COSTRUTTORI
	public Animale(int eta, Genere sesso, float peso) throws BadInputException {
		this.id = "A" + (++count);
		if(eta < 0 || peso <= 0) throw new BadInputException();
		this.eta = eta;
		this.peso = peso;
		this.sesso = sesso;
		this.vivo = true;
	}
	
	public Animale(Genere sesso, float peso) throws BadInputException {
		this(0, sesso, peso);
	}

	//METODI
	public int getEta() {
		return eta;
	}
	public String getId() {
		return id;
	}
	public float getPeso() {
		return peso;
	}
	public Genere getSesso() {
		return sesso;
	}
	public boolean isVivo() {
		return vivo;
	}
	public void setPeso(float p) {
		this.peso = p;
	}
	public void invecchia() {
		this.eta++;
	}
	public void setVivo(boolean vivo) {
		this.vivo = vivo;
	}
	@Override
	public String toString() {
		return id + "\n   mesi = " + eta + "\n   peso = " + peso  + " kg" + "\n   sesso = " + sesso + "\n   vivo = " + (vivo ? "sì" : "no");
	}
	//ordinamento di default dipende dal solo id univoco
	@Override
	public int compareTo(Animale a) {
		return this.id.compareTo(a.id);
	}
}
